src/QsCompiler/CompilationManager/CompilationUnit.cs

```diff
@@ -24,117 +25,114 @@ public class References
                 IEnumerable<TypeDeclarationHeader> types = null)
             {
-                this.Callables = callables?.ToImmutableArray() ?? ImmutableArray<CallableDeclarationHeader>.Empty;
-                this.Specializations = specs?.ToImmutableArray() ?? ImmutableArray<SpecializationDeclarationHeader>.Empty;
-                this.Types = types?.ToImmutableArray() ?? ImmutableArray<TypeDeclarationHeader>.Empty; 
+                NonNullable<string> SourceOr(NonNullable<string> origSource) => NonNullable<string>.New(source ?? origSource.Value);
+                this.Types = types?.Select(t => t.FromSource(SourceOr(t.SourceFile))).ToImmutableArray() ?? ImmutableArray<TypeDeclarationHeader>.Empty;
+                this.Callables = callables?.Select(c => c.FromSource(SourceOr(c.SourceFile))).ToImmutableArray() ?? ImmutableArray<CallableDeclarationHeader>.Empty;
+                this.Specializations = specs?.Select(s => (s.Item1.FromSource(SourceOr(s.Item1.SourceFile)), s.Item2 ?? SpecializationImplementation.External)).ToImmutableArray() 
+                    ?? ImmutableArray<(SpecializationDeclarationHeader, SpecializationImplementation)>.Empty;
             }
+
+            // TODO: keep indirect references? 
+            // -> would need to make sure it's fine to pull in the same reference multiple times though...
+            internal Headers(NonNullable<string> source, IEnumerable<QsNamespace> syntaxTree) : this (
+                source.Value, 
+                syntaxTree.Callables().Where(c => c.SourceFile.Value.EndsWith(".qs")).Select(CallableDeclarationHeader.New),
+                syntaxTree.Specializations().Where(c => c.SourceFile.Value.EndsWith(".qs")).Select(s => (SpecializationDeclarationHeader.New(s), s.Implementation)), 
+                syntaxTree.Types().Where(c => c.SourceFile.Value.EndsWith(".qs")).Select(TypeDeclarationHeader.New))
+            { }
+
+            internal Headers(NonNullable<string> source, IEnumerable<(string, string)> attributes) : this(
+                source.Value,
+                References.CallableHeaders(attributes), 
+                References.SpecializationHeaders(attributes).Select(h => (h, (SpecializationImplementation)null)), 
+                References.TypeHeaders(attributes))
+            { }
         }
 
+        private static Func<(string, string), string> IsDeclaration(string declarationType) => (attribute) =>
+        {
+            var (typeName, serialization) = attribute;
+            if (!typeName.Equals(declarationType, StringComparison.InvariantCultureIgnoreCase)) return null;
+            return serialization;
+        };
+
+        private static IEnumerable<CallableDeclarationHeader> CallableHeaders(IEnumerable<(string, string)> attributes) =>
+            attributes.Select(IsDeclaration("CallableDeclarationAttribute")).Where(v => v != null)
+                .Select(CallableDeclarationHeader.FromJson).Select(built => built.Item2);
```

Comment:
One note that might be irrelevant but isn't really well documented: most of the LINQ extension methods have the feature that they evaluate lazily. In many cases this is a significant benefit, but it can be a real problem if there's any possibility of the underlying collection getting modified before you actually complete the enumeration (by turning it into an ImmutableArray, for instance). So for these three static methods, for instance, you'd have to be careful to complete the enumeration before you change the argument. If these are only used in the constructor above, you're fine because the root constructor always completes the enumeration, but I'm not sure where else you use these.

Reply:
Yes, I know. They are indeed private and only used in a save context, so I am omitting the unnecessary array creation.